Patch by MSP-Greg
Attempts to fix issues when run on 'busy' CI systems
diff --git a/bootstraptest/runner.rb b/bootstraptest/runner.rb
index 2cd91ffd07..2728905040 100755
--- a/bootstraptest/runner.rb
+++ b/bootstraptest/runner.rb
@@ -402,9 +402,15 @@ def assert_finish(timeout_seconds, testsrc, message = '')
       end
       if io.respond_to?(:read_nonblock)
         if IO.select([io], nil, nil, diff)
+          tries = 0
           begin
             io.read_nonblock(1024)
-          rescue Errno::EAGAIN, IO::WaitReadable, EOFError
+          rescue IO::WaitReadable
+            IO.select([io])
+            tries += 1
+            break if tries > 3
+            retry
+          rescue Errno::EAGAIN, EOFError
             break
           end while true
         end
diff --git a/bootstraptest/test_io.rb b/bootstraptest/test_io.rb
index 89c00d0b88..6a240b3cd2 100644
--- a/bootstraptest/test_io.rb
+++ b/bootstraptest/test_io.rb
@@ -1,4 +1,4 @@
-assert_finish 5, %q{
+assert_finish 6, %q{
   r, w = IO.pipe
   t1 = Thread.new { r.sysread(1) }
   t2 = Thread.new { r.sysread(1) }
diff --git a/bootstraptest/test_thread.rb b/bootstraptest/test_thread.rb
index 38a55ff229..6a91a35fcd 100644
--- a/bootstraptest/test_thread.rb
+++ b/bootstraptest/test_thread.rb
@@ -335,7 +335,7 @@ def m
 
 assert_normal_exit %q{
   Thread.new("foo", &Object.method(:class_eval)).join
-}, '[ruby-dev:34128]'
+}, '[ruby-dev:34128]', 'SIGSEGV'
 
 assert_equal 'ok', %q{
   begin
